### DISCUSSION TITLE
Confused about some concepts in our data structures

### DISCUSSION THREAD

**Discussion by haruming** [mspass-team/mspass#113] [Q&A]
Hi all, hopefully, I didn't start a new discussion in the wrong category:) I have been looking into our codebase these days and trying to understand our basic data structures in mspass so that I could better understand some more complicated ones. However, I feel a little confused about some concepts in our data structures.

1. How to understand the **shift** method in our BasicTimeSeries and are there any examples in seismology that we need to shift a reference time t0? Also, I found that in the implementation, the **shift** method didn't use the parameter **dt**, so I am wondering is that a typo?
```
void BasicTimeSeries::shift(double dt)
{
    try {
        double oldt0shift=t0shift;
        this->rtoa();
        this->ator(oldt0shift+mdt);
    } catch(...) {
        throw;
    };
}
```

2. For our mspass::utility::dmatrix in seismogram, my understanding is we store our matrix values in the 1-dimension vector **std::vector<double> ary** and we could access the value through row_index and col_index, normally it will be ary[col_index+(ncc)*(row_index)]. However, in our implementation of the operator(), it accesses the value by ary[row_index+(nrr)*(col_index)], so I feel a little confused by the logic here.
```
double dmatrix::operator()(const size_t rowindex, const size_t colindex) const
{
  int out_of_range=0;
  if (rowindex>=nrr) out_of_range=1;
  if (rowindex<0) out_of_range=1;
  if (colindex>=ncc) out_of_range=1;
  if (colindex<0) out_of_range=1;
  if (out_of_range)
	throw dmatrix_index_error(nrr,ncc,rowindex,colindex);
  double result=ary[rowindex+(nrr)*(colindex)];
  return result;
}
```

After some research I think question 2 is because data in the matrix is ordered in a Fortran way, but why?

**Comment by wangyinz** [comment]
Well, I guess I would categorize this as internal discussions, which we should resolve with email or slack. But, I do think your questions are really good. For the shift function, yeah, that is confusing to me, too. Maybe there is a bug here? @pavlis 

For question 2, I am glad that you found the answer by yourself so quickly. Yes, we are using the fortran order here. I guess there is a historical reason to it that @pavlis could probably explain in great details.

**Comment by pavlis** [comment]
Good questions:  Always the sign of a good student when they ask questions that good.  

First, the shift method.   That question is timely because I've been working on the documentation and a couple days ago finished a separate page for the documentation with an expanded discussion of the concepts behind this design.   I'm not quite read to push the changes yet because I'm reworking all of the existing user's manual pages to fix inconsistencies from changes made since the originals were developed.  I don't think I have your email address.  Suspect there is a way to get it from github, but I'm ignorant and too lazy to do a google search to find out.  Maybe Ian can send me your email if you don't have mine - I don't want to put an email address directly in a public page like this.  

An example of how that particular method might be used is this.  A person might download a set of data from IRIS for a set of earthquakes.   Suppose they downloaded time windows that start at the time of the earthquake and called the *ator* method using the unique time seismologists call the "origin time".   Suppose *d* is a TimeSeries constructed that way.  If you ran this line:  `t=d.time(0)` the time return, t, would be 0.0 because after ator is called the data are in "relative" time with 0 being the origin time.  How would we use shift?  It isn't common, but here is the example.  Suppose we wanted to focus on data around some seismic "phase" which means a burst of energy with a predictable timing.  The type example would be "P" and "S" phases which you likely learned in elementary school, especially in a country like China prone to deadly earthquakes.  P and S are phases with a time that can be measured or predicted from earth models.  The later is most relevant here as one could use an earth model to predict the P wave arrival time.   An example is [this function](https://docs.obspy.org/packages/obspy.taup.html) in obspy.   Let `Tp` be the "travel time" return by the taup calculator.  Then a user can shift 0 for d to the predicted time of P with this command:  `d.shift(Tp)`.  

You have answered the second question yourself and I concur with Ian that I'm happy to see you do that.  The motivation behind that choice is described to some extent in our User's Manual on [this page](https://wangyinz.github.io/mspass/user_manual/data_object_design_concepts.html) Look especially at the section there titled "Scalar versus 3C Data".   A more subtle issue to educate you is that one thing I always thought Kernighan and Ritchie really botched when the created C was the way the defined arrays in the language.   Don't know how well you know C or Fortran but they store arrays in completely different ways.  

In FORTRAN an array is just a block of contiguous memory.  e.g. a 3x2 double matrix declared in FORTRAN would be stored in a contiguous block of 6 double (8 byte) blocks.  Simple and easy to work with, and the example you refer to above shows why - the indexing can be computed so A(i,j) (the FORTRAN form) is just syntactic sugar for A(i+nrr*j).  

Would that things would be that simple in C.   To even create a 3x2 array in C requires this type of obnoxious construct:
```
double **A;
for(auto i=0;i<3;++i) A[i]=(double)calloc(2,sizeof(double));
```
Once you do that in C you can reference say 1,2 as `A[1][2]`.  You often see sources simply this obnoxious thing by saying Fortran is row order and C is column order, but that can be very misleading.   As the above construct shows you cannot assure in C that the pieces are in contiguous memory.   Probably more significant is that fact that in C destruction of a 2 D array in C is subject to two serious problems:  (1) it takes more computing time because the system has to call free on each row of the matrix, and (2) (worst) is it is prone to memory leaks if a person using the data isn't aware of this C oddity.  

That was a long lesson, but it was a good question about a subtle property of C that in my opinion deserves an English word you likely have to look up in a dictionary:  abomination.

**Reply by wangyinz** [reply]
Hey Gary, I do think we have a typo there in:

```
void BasicTimeSeries::shift(double dt)
{
    try {
        double oldt0shift=t0shift;
        this->rtoa();
        this->ator(oldt0shift+mdt);
    } catch(...) {
        throw;
    };
}
```
The `ator` line should be `this->ator(oldt0shift+dt);` instead.

**Reply by pavlis** [reply]
You are absolutely right that is a mistake.  I completely understand where it came from too.  Recall when we changed the api to add and getter and setter for the sample interval we created the dt() and set_dt() methods in BasicTimeSeries.  Because of C++ symbol restrictions that required the old variable called "dt" to be changed everywhere to "mdt".   I obvious mess up editing that method.  

Are you going to change this or should I?  As the keeper of github it would probably be easier for you to make this change than doing a pull request for a single bug fix like that.

**Reply by wangyinz** [reply]
yes I can do that, but this is actually a good example of why a good test coverage is important. It could have been captured by a test, but I guess we never have one for time shift.

**Reply by wangyinz** [reply]
This is fixed in https://github.com/wangyinz/mspass/commit/cd34b13b9b31dada5549a4527ade5e4275c25e15
